File: modules/core/src/main/java/kr/entree/enderwand/command/completer/TabCompleter.java
package kr.entree.enderwand.command.completer;

import kr.entree.enderwand.command.CommandContext;
import kr.entree.enderwand.command.sender.Sender;

import java.util.List;

/**
 * Created by JunHyung Lim on 2020-02-04
 */
public interface TabCompleter<S extends Sender> {
    List<String> tabComplete(CommandContext<S> ctx);
}

File: modules/core/src/main/kotlin/kr/entree/enderwand/command/tabcompleter/TabCompleter.kt
package kr.entree.enderwand.command.tabcompleter

import kr.entree.enderwand.collection.Reader
import kr.entree.enderwand.command.CommandContext
import kr.entree.enderwand.command.sender.Sender

/**
 * Created by JunHyung Lim on 2019-12-18
 */
typealias FunctionalTabCompleter<S> = (CommandContext<S, Reader<String>>) -> List<String>

fun <S : Sender> TabCompleter<S>.tabComplete(sender: S, args: Reader<String>) =
    tabComplete((CommandContext(sender, args)))

interface TabCompleter<S : Sender> {
    fun tabComplete(
        ctx: CommandContext<S, Reader<String>>
    ): List<String>
}